README.md
# Nextjs + Cordova + Framework7

Use this package if you want to configure an application running Next, Cordova, Framework 7.
Real-time development mode on an emulator device, phone, browser.
This package allows you to test in real time, write code together with plugins, without resorting to pre-installing the application. And only after changing /refining the plugin code - recompilation of the application will be required. Therefore, you can write code calmly, with ready-made plugins in real time on the emulator with a quick page reload in the next dev development mode.
Also, this package has a full extension to use other framework packages without resorting to global changes. It's enough to remove some details that you don't need.

This package uses [next-cordova-static](https://www.npmjs.com/package/next-cordova-static)

You can help the project:

_USDT TRC20: THyUxnKfShQ9YkXTEUwNvYZiWgvrR9PCM1_


# Install 
```
npm install nextjs-cordova-framework7 -g
```

or

```
yarn global add nextjs-cordova-framework7
```
# Package.json
Parameter | Default | Description
-- | -- | --
`ncfParams.port.http` | 9090 | Development port
`ncfParams.port.https` | 9091 | Development port
`ncfParams.usesCleartextTraffic` | true | For android platform usesCleartextTraffic

# Available commands
Command | Description
-- | --
`create-ncf-app` | Creates an application of the current directory, ready to work
`ncf dev` | Runs two development servers, on http and https
`ncf emulate [platform] [option]` | Run the build for the emulator
`ncf run [platform] [option]` | Run the build for the device

&nbsp;

Platform | Description
-- | --
`android` | &#9745;
`ios` | &#9744; (Work is underway)

&nbsp;

Option | Description
-- | --
`dev` | Real-time development. 

# Steps required

Run the command that will create in the current directory all the necessary tools to work with the application:
```
create-ncf-app
```


# Build an application for development
```
ncf emulate android
```
This is where the static build is launched, already a full-fledged application in development mode. This mode is needed in order to show, for example, a manufactured application to the customer, and test it all together.


# Developing an application
Developing an application on port 9090 (change if necessary)
```
ncf emulate android dev && ncf dev
```
Here you will run a command to build the application in development mode, i.e. the application will look towards your computer, and thus you will be able to launch a local web server that will show the content in the application. You can quickly, conveniently, with all plugins, enjoy the development of the application.



# General information

You can add all other commands directly to your package.

Commands that you can add:

1. [Cordova](https://cordova.apache.org/docs/en/latest/)

2. [NextJS](https://nextjs.org/docs/getting-started)

3. Use the documentation for building the application [Framework7](https://framework7.io/react/)

Do not try to create an application again using the "create next-app" or "cordova create" commands, everything you need is already present in this package, and you only need to work on your application.

Everything else - You can use it as if you used cordova, nextjs, framework 7 separately

# Example

Example of working with a package:

1. Creating an application
```
create-ncf-app
```

2. Adding a platform
```
cordova platform add android
```

3. We use the real-time development mode (in this case, an emulator)
```
ncf emulate android dev
```

4. Start the development server:
```
ncf dev
```

5. Launch the app on your phone or emulator

6. Edit the code in real time, with all the features of the emulator/phone, as well as with all available plugins.

7. Viewing the device [chrome://inspect/#devices](chrome://inspect/#devices)

CHANGELOG.md
# Change Log

### 1.2.8
* Added package local-ssl-proxy
* HTTPS development
* Added port indication for http and https
* Added specifying ports for http and https in ncfParams.port

### 1.2.6
* Added usesCleartextTraffic
* Update description
* fix minor

### 1.2.3
* fix minor

### 1.2.2
* Big changes

### 0.1.5 
* Add Release Notes
* Update package

### 0.1.3 
* Added deviceready

### 0.1.1 
* Public folder added

### 0.0.9 
* Added a test catalog template

### 0.0.7 
* Add "create-ncf-app"

### 0.0.4 
* Minor fixes 

### 0.0.3 
* Add build && export

### 0.0.2 
* Minor fixes

### 0.0.1 
* Create app

src/function.js
const shelljs = require('shelljs');
const fse = require('fs-extra');

function shelljsExec(command) {
    return new Promise(function (resolve, reject) {

        shelljs.exec(command, {}, function (code, value, error) {

            if (error) {

                return resolve(error);
                
            }

            resolve(value);
        });

    });
}

function editConfig ({ action, method, cwd, url }) {
    return new Promise(async function (resolve, reject) {

        action = action ?? null;
        url = url ?? null;
        cwd = cwd ?? null;
        method = method ?? false;

        if (!action) {

            reject('Action not specified');

        }

        if (!cwd) {

            reject('cwd not specified');

        }

        if (action == 'content' && !url) {

            reject('Url not specified');

        }

        const getFile = await fse.promises.readFile(cwd + '/config.xml');

        let data = getFile.toString();

        if (action == 'content') {
            
            console.log('Setting the url: ' + url);

            data = data.replace(/(<content [\S\s]*?src=")[^"]+("[\S\s]*?>)/gmi, '<content src="' + url + '" />');

        } else if (action == 'usesCleartextTraffic') {


            if (method) {

                if (!/<widget.+xmlns:android/i.test(data)) {

                    console.log('Added config.xml <widget> -> @xmlns:android="http://schemas.android.com/apk/res/android"');

                    data = data.replace(/<widget(.+?)>/gmi, '<widget$1 xmlns:android="http://schemas.android.com/apk/res/android">');

                }

                if (!/usesCleartextTraffic/i.test(data)) {

                    console.log('Added android:usesCleartextTraffic="true"');
                    
                    data = data.replace(/<platform\sname="android"([\s\S]+?)<\/platform>/gmi, '<platform name="android"$1<edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application"><application android:usesCleartextTraffic="true" /></edit-config><\/platform>');

                }

            } else {

                console.log('Remove config.xml <widget> -> @xmlns:android="http://schemas.android.com/apk/res/android"');

                data = data.replace(/<widget(.+?)\sxmlns:android=".+">/gmi, '<widget$1>');

                console.log('Remove android:usesCleartextTraffic="true"');

                data = data.replace(/<edit-config[\s\S]+usesCleartextTraffic[\s\S]+<\/edit-config>/gmi, '');

            }

        }

        await fse.promises.writeFile(cwd + '/config.xml', data);

        resolve(true);

    });
}

module.exports = {
    shelljsExec,
    editConfig
}

template/pages/_app.js
import Framework7 from 'framework7/bundle';
import Framework7React, { f7ready, App, View } from 'framework7-react';
Framework7.use(Framework7React);
import '../node_modules/framework7/framework7-bundle.min.css';
import '../public/static/css/main.css';
import PanelLeft from '../components/PanelLeft.js';
import { useRouter } from 'next/router';
import Head from 'next/head';
import Routes from '../components/Routes';
import { useEffect, useState } from 'react';

function CordovaApp({ Component, pageProps }) {

    const router = useRouter();
    const [app, setApp] = useState(null);
    const [isDeviceready, setIsDeviceready] = useState(false);

    useEffect(() => {
        f7ready((f7) => {
            setApp(f7);
        });
    }, []);

    useEffect(() => {
        if (typeof document == 'undefined') return;
        document.addEventListener('deviceready', function () {
            setIsDeviceready(true);
        });
    });
    
    return <>
        <Head>
            <title>My page title</title>
            <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no, minimal-ui, viewport-fit=cover" />
            <meta name="format-detection" content="telephone=no" />
		    <link rel="icon" href="/favicon.ico" type="image/x-icon" />
            {(() => {
                if (app) {
                    if (
                        app.device.android == true ||
                        app.device.ios == true
                    ) {
                        return <>
                            <script src="cordova.js"></script>
                        </>;
                    }
                }
            })()}
        </Head>
        <App 
            theme="ios" 
            url={(process.env.NEXT_PUBLIC_HOST + router.asPath)} 
            routes={Routes}
        >
            <PanelLeft />
            <View url="/" className="view-main view-init">
                <Component initialPage {...pageProps} />
            </View>
        </App>
    </>;
}

export default CordovaApp;

src/cli.js
const fse = require('fs-extra');
const { shelljsExec, editConfig } = require('./function');


module.exports = async function (Args) {
    try {

        // args
        const inputArgs = Args || process.argv;

        // command
        const command1 = inputArgs[2] ?? null;// dev, emulate, run
        const command2 = inputArgs[3] ?? null;// android, ios
        const command3 = inputArgs[4] ?? null;// dev

        // work catalog
        const cwd = process.cwd();
        
        // package
        const package = require(cwd + '/package.json');

        // port http
        const portHttp = package?.ncfParams?.port?.http ?? 9090;

        // port https
        const portHttps = package?.ncfParams?.port?.https ?? 9091;

        // usesCleartextTraffic
        const usesCleartextTraffic = package?.ncfParams?.usesCleartextTraffic ?? true;

        // dev
        if (command1 == 'dev') {

            shelljsExec("npx next dev -p " + portHttp);

            shelljsExec("npx local-ssl-proxy --source " + portHttps + " --target " + portHttp);

        // emulate or run
        } else if (command1 == 'emulate' || command1 == 'run') {

            // android
            if (command2 == 'android') {

                // delete dir 'public/plugins'
                if (fse.existsSync(cwd + '/public/plugins')) {

                    console.log('Remove "/public/plugins"');

                    fse.removeSync(cwd + '/public/plugins');
                }

                // delete file 'public/cordova_plugins.js'
                if (fse.existsSync(cwd + '/public/cordova_plugins.js')) {

                    console.log('Remove "/public/cordova_plugins.js"');

                    fse.removeSync(cwd + '/public/cordova_plugins.js');
                }

                // delete file 'public/cordova.js'
                if (fse.existsSync(cwd + '/public/cordova.js')) {

                    console.log('Remove "/public/cordova.js"');

                    fse.removeSync(cwd + '/public/cordova.js');
                }

                // command3 == 'dev'
                if (command3 == 'dev') {

                    // copy dir 'platforms/android/platform_www' -> 'public'
                    if (fse.existsSync(cwd + '/platforms/android/platform_www')) {

                        console.log('Copy dir "/platforms/android/platform_www" -> "public"');

                        fse.copySync(cwd + '/platforms/android/platform_www', 'public');

                    }

                    // rename config.xml: <content src="https://localhost" /> -> <content src="http://10.0.2.2:9090/" />
                    await editConfig({
                        cwd,
                        action: 'content',
                        url: 'https://10.0.2.2:' + portHttps
                    });

                }

                // usesCleartextTraffic
                if (usesCleartextTraffic) {

                    // add usesCleartextTraffic
                    await editConfig({
                        cwd,
                        action: 'usesCleartextTraffic',
                        method: true
                    });

                }

                console.log('Run next-cordova-static');

                await shelljsExec("npx next-cordova-static");

                console.log('Run cordova ' + command1 + ' android');

                await shelljsExec("npx cordova " + command1 + " android");

                // command3 == 'dev'
                if (command3 == 'dev') {

                    // rename config.xml: <content src="http://10.0.2.2:9090/" /> -> <content src="https://localhost" />
                    await editConfig({
                        cwd,
                        action: 'content',
                        url: 'https://localhost'
                    });

                }

                // usesCleartextTraffic
                if (usesCleartextTraffic) {

                    // del usesCleartextTraffic
                    await editConfig({
                        cwd,
                        action: 'usesCleartextTraffic',
                        method: false
                    });

                }

            } else {

                console.log('Error: Enter the platform');
                
            }

        } else {

            console.log('Error: Enter the command');

        }
        
    } catch (error) {

        console.log('Fatal error: ', error);

    }
}



